### PR TITLE
Copy rebar.app before bootstrapping

### PR DESCRIPTION
**Pull request by big-r81**
Calling 'make clean' from CouchDB wipes all ebin-directories in the source tree. 
Afterwards it is not possible to build rebar again. 

Modifying the bootstrap script and moving 'rebar.app' one directory-level up.
Before building, copy rebar.app into ebin/.

### CODE REVIEW

**Comment by jaydoane** [comment]
I ran into this issue today, so appreciate that it could be improved!

However, moving rebar.app file to the top level looks weird, even though it seems to build that way.

What if we moved it to src/rebar.app and then in bootstrap ensure ebin exists then copy it there from src/rebar.app?

**Comment by big-r81** [comment]
I have an open PR on the main repo too. Can you test this PR with the combination of https://github.com/apache/couchdb/pull/4340

**Comment by jaydoane** [comment]
According to [the docs](https://www.erlang.org/doc/man/app.html#file-syntax):

> The file is to be located in directory ebin for the application

Maybe a better solution would be to fix `make clean`? What if we changed [this line](https://github.com/apache/couchdb/blob/main/Makefile#L442) to e.g.
```
	@rm -f src/*/ebin/*.beam
```
?

**Comment by big-r81** [comment]
> According to [the docs](https://www.erlang.org/doc/man/app.html#file-syntax):
> 
> > The file is to be located in directory ebin for the application

Okay, then let's follow the docs.

> Maybe a better solution would be to fix `make clean`? What if we changed [this line](https://github.com/apache/couchdb/blob/main/Makefile#L442) to e.g.
> 
> ```
> 	@rm -f src/*/ebin/*.beam
> ```

I think about that too, but then the `ebin` dirs are not removed, but empty.
I will update the PR, that the `rebar.app` is copied to `ebin` before building.

**Comment by jaydoane** [comment]
> I think about that too, but then the ebin dirs are not removed, but empty.

I thought more about this as well. They would not quite be empty, but still contain the .app files generated by rebar from the .app.src files, which I think could be a problem since I think we'd want `make clean` to remove those as well.

**Comment by big-r81** [comment]
I think the PR is ready. On bootstrapping it checks if beim exists (create it if not) and copy the rebar.app into it. In combination with https://github.com/apache/couchdb/pull/4340 it’s now possible to rebuild rebar if missing in bin/

**Comment by jaydoane** [comment]
It appears that you've added a new copy of ebin/rebar.app in the top level while leaving the original in place.

Did you mean to do e.g. `git mv ebin/rebar.app .` instead?

**Comment by big-r81** [comment]
Oh yes, it's a leftover of the changes. Since we now always copy rebar.app when building, it is of course sufficient to leave it in the top-level directory.

File: bootstrap [hunk 1]
```diff
@@ -49,6 +49,12 @@ main(Args) ->
                           false -> undefined
                       end,
 
+    %% Ensure, that the ebin folder exists after "make clean"
+    ok = filelib:ensure_dir("ebin/"),
+
+    %% Copy rebar.app into ebin
+    {ok, _} = file:copy("rebar.app", "ebin/rebar.app"),
```
Comment (jaydoane) [inline bootstrap:55]: maybe we can add something like e.g. ?
```
    %% CouchDB's `make clean` clobbers all ebin directories, so protect rebar.app by
    %% moving it out of harm's way, and copying it into ebin before building.
```

**Review by jaydoane — APPROVED** [review]
Tested locally, and I'm able to build successfully after running `make clean` in CouchDB.

Please rebase before merging.

Hmm, I just noticed that ebin/rebar.app isn't ignored by git. I wonder if we want to add it to .gitignore or change the existing ebin/*.beam line?

File: bootstrap (hunk 1 above)
Comment (big-r81) [reply bootstrap:55]: Added your suggestion and added ebin/ to .gitignore.

**Review by jaydoane — APPROVED** [review]
Nice fix! Thanks for making it easier to hack on this obsolete software ;)

**Comment by big-r81** [comment]
Maybe a final tribute before moving to rebar3... 😉

**Comment by jaydoane** [comment]
> Maybe a final tribute before moving to rebar3...

We've tried migrating several times, but maybe we'll get there :)

**Pull request merged**